packages/redux/src/checkout/reducer.ts

```diff
@@ -172,20 +173,72 @@ const convertCheckoutOrder = (
     !!get(mergedState, `checkoutOrders[${result}].clickAndCollect`) &&
     !get(entities, `checkoutOrders[${result}].clickAndCollect`);
 
-  if (!shouldResetClickAndCollect) {
-    return mergedState;
-  }
+  const mergedShippingAddressState = get(
+    mergedState,
+    `checkoutOrders[${result}].shippingAddress.state`,
+  );
+  const receivedShippingAddressState = get(
+    entities,
+    `checkoutOrders[${result}].shippingAddress.state`,
+  );
+  const shouldReplaceShippingState = !isEqual(
+    mergedShippingAddressState,
+    receivedShippingAddressState,
+  );
+
+  const mergedBillingAddressState = get(
+    mergedState,
+    `checkoutOrders[${result}].billingAddress.state`,
+  );
+  const receivedBillingAddressState = get(
+    entities,
+    `checkoutOrders[${result}].billingAddress.state`,
+  );
+  const shouldReplaceBillingState = !isEqual(
+    mergedBillingAddressState,
+    receivedBillingAddressState,
+  );
+
+  return produce(mergedState, draftState => {
+    if (!draftState.checkoutOrders[result]) {
+      return;
+    }
 
-  return {
-    ...mergedState,
-    checkoutOrders: {
-      ...mergedState.checkoutOrders,
-      [result]: {
-        collectpoints: mergedState.checkoutOrders[result]?.collectpoints,
+    if (
+      !shouldResetClickAndCollect &&
+      !shouldReplaceBillingState &&
+      !shouldReplaceShippingState
+    ) {
+      return;
+    }
+
+    if (shouldResetClickAndCollect) {
+      draftState.checkoutOrders[result] = {
+        collectpoints: draftState.checkoutOrders[result]?.collectpoints,
         ...entities?.checkoutOrders[result],
-      },
-    },
-  };
+      };
+    }
+
+    if (shouldReplaceShippingState) {
+      draftState.checkoutOrders[result] = {
```

Comment:
It is not necessary to reassign the checkout order in draftState since immer will take care of that.
In this case what you wanted is to do something like this:

```js
draftState.checkoutOrders[result]?.shippingAddress.state = entities.checkoutOrders[result]?.shippingAddress?.state;
```

Then, when the call to `produce` finishes, immer will check which properties have changed and will generate a new state object with new references only for those properties that changed.

If, instead, you meant to replace the whole shipping address if `state` has changed, then you will need to adapt to something like this:

```js
draftState.checkoutOrders[result]?.shippingAddress= entities.checkoutOrders[result]?.shippingAddress;
```
So, in summary, you only need to treat the `draftState` as a mutable object and immer will take care of generating the object will new references where necessary.